### ISSUE TITLE
HtmlAttribute.Value should use HtmlEntity.DeEntitize when setting _value

### ISSUE THREAD

**Issue by drbsoftware** [zzzprojects/html-agility-pack#19]
The Value of attribute data-foo="&amp;quot;Hello&amp;quot;" should be "Hello", not &amp;quot;Hello&amp;quot;
I think the fix is:
```        public string Value
        {
            get
            {
                if (_value == null)
                {
                    _value = HtmlEntity.DeEntitize(ownerdocument.Text.Substring(_valuestartindex, _valuelength));
                }
                return _value;
            }
            set
            {
                _value = HtmlEntity.DeEntitize(value);
                if (_ownernode != null)
                {
                    _ownernode._innerchanged = true;
                    _ownernode._outerchanged = true;
                }
            }
        }
```

**Comment by JonathanMagnan** [comment]
Hello @drbsoftware ,

Thank you, my programmer will look at this issue during the week.

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Hello @drbsoftware ,

The v1.5.0-beta92 now use your fix.

Let me know if that's correctly working.

Best Regards,

Jonathan

**Comment by yfayerman** [comment]
HtmlEntity.DeEntitize method should not be called when setting HtmlAttribute.Value as it must be properly encoded, see [this](https://stackoverflow.com/questions/13246540/html-and-attribute-encoding).

As for getting the value, the property should, IMHO, return its original encoded value, and it's up to the caller whether to decode it or not.

**Comment by JonathanMagnan** [comment]
Hello @yfayerman ,

Thank you for reporting. I will have my developer look at it and check if we must do it differently.

Best Regards,

Jonathan

**Comment by drbsoftware** [comment]
8.1.2.3 Attributes in [https://www.w3.org/TR/html5/syntax.html](https://www.w3.org/TR/html5/syntax.html) would seem to suggest that the parsing is the same for attributes and for HTML text.

In our case, we are using [custom data-* attributes](https://www.w3.org/TR/html5/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes).  The data is stored by JavaScript using the Chromium DOM.  We retrieve the attribute contents using HTML Agility Pack.  We expect the data retrieved to be the same as the data stored.

Thanks
David

**Comment by JonathanMagnan** [comment]
Hello @yfayerman  ,

The setter does not longer use DeEntitize.

The getter however still use it in order to solve @drbsoftware first issue. To mimic what a browser does. By example `data-foo="&quot;Hello&quot;"` are parsed as `data-foo="\"Hello\""` in the code source of a browser.

Please @yfayerman , @drbsoftware , let me know if we can close this issue or there is much work to do on it.

Best Regards,

Jonathan

**Comment by drbsoftware** [comment]
Jonathan, I think that looks right.
Thanks
David

**Comment by yfayerman** [comment]
@JonathanMagnan, that's fine with me.

**Comment by JonathanMagnan** [comment]
Closing Comment: Fix confirmed

**Comment by JonathanMagnan** [comment]
Hello @drbsoftware , @yfayerman ,

Finally, we had to rollback this fix in v1.5.1. @yfayerman is right, it should be up to the caller instead to DeEntitize or not the value. We cannot automatically DeEntitize it otherwise we will break some existing code like in the issue #32.

However, to make it easier, we added a new property `DeEntitizeValue` which return the expected value `"Hello"` if you want it decoded.

Best Regards,

Jonathan

**Comment by drbsoftware** [comment]
Hi Jonathan,
It seems to me that the issues here and in #32 are:
- What is written out should be what is read in (unless modified by options or by external software ).
- What is accessed by .NET software should be equivalent to what is accessed by JavaScript on the DOM.

I think that the problem encountered by #32 is simply that the fix to #19 was incomplete. HtmlNode.WriteAttribute also needs to be modified to use Entitize.

David
BTW: Thank you for your efforts in keeping this library maintained.

**Comment by JonathanMagnan** [comment]
Hello @drbsoftware ,

Thank you ;) We try to do our best!

Using the WebBrowser and some other HTML parser, you are right. The value should return `"Hello"`

However, the SetAttribute (using a WebBrowser) doesn't DeEntitize the value. `&quot;` will be instead modified as `&amp;quot;` instead in the HTML so the value return `&quot;`

Until we rewrite this library, I believe the best is to stay with the current behavior since a `DeEntitizeValue` property has been added. If you need, we can, however, add an option.

Way too much people are currently using the library with this behavior, so staying backward compatible is maybe more important in this case than doing the right thing.

Best Regards,

Jonathan
